### PR TITLE
[FLOSS T3] docs: Improve Git Documentation

### PR DESCRIPTION
**Pull request by Bujuhu**
Co-authored-by: Dynamichost96 <60710121+Dynamichost96@users.noreply.github.com>

We looked at, which parts of the git flow we struggled with during our contribution to libelektra and tried to add them to the existing GIT.md guide

We worked together on this document

## Basics

- [ ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - [ ] add a line in `doc/news/_preparation_next_release.md`
  - [ ] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [ ] The PR is rebased with current master.

## Checklist

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation (see [Documentation Guidelines](https://www.libelektra.org/devgettingstarted/documentation))
- [ ] I fixed all affected decisions (see [Decision Process](https://www.libelektra.org/decisions/decision-process))
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://www.libelektra.org/devgettingstarted/coding))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned [every code](/.reuse/dep5) not directly written by me in [reuse syntax](https://reuse.software/)

## Review

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

- [ ] Add the "work in progress" label if you do not want the PR to be reviewed yet.
- [ ] Add the "ready to merge" label **if the basics are fulfilled** and no further pushes are planned by you.

### CODE REVIEW

File: doc/GIT.md [hunk 1]
```diff
@@ -115,3 +140,17 @@ git push orign master:master
...
+
+To achieve this, make sure you have your master branch synced up. Then switch to the feature branch you want to update using `git checkout <feature-branch-name>` and then type `git rebase master`. Sometimes, this will result in a merge conflict.
+
+If you already have pushed changes from your feature branch to a remote branch, you'll need to either overwrite it using `git push --force` or push to a new remote branch
+
+### Resolving merge conflicts
+
+To resolve these edit the files that need to be merged manually. You can check which files need to be merged using `git status`. After you are done merging a file manually, you can use `git add <path-to-file>` and when you have merged all files, continue with `git rebase --continue`.
+
+> **Note**: Keep in mind that doing manual merges within a rebase are destructive. If you accidentally remove changes you've done in your feature branch, they will be gone forever. If you're new to manually merging files in git, it's a good idea to create a new branch from your feature branch using `git checkout -b <new-branch-name>`. When you're stuck on a manual merge, you can also always abort the rebase using `git rebase --abort`
```
Comment (0x6178656c) [inline doc/GIT.md:156]: This is not entirely true. By default Git will not remove previous commits immediately, even if they are no longer reachable from any of the branches.

Unreachable commits are only deleted when the garbage collector is invoked. (The documentation is not very specific about when this occurs.)

In any case there is `git-reflog` which allows one to retrieve previous commits even if they are no longer reachable from any branch; provided they have not been deleted yet by the garbage collector.

* <https://git-scm.com/docs/git-reflog/2.39.0>
* <https://git-scm.com/docs/git-gc>

(Basically anything you have committed is fine to overwrite using Git commands. The only dangerous thing is doing `git reset` when having changes in the working tree/index which have not been committed at any point in time.)

**Review by 0x6178656c — COMMENTED** [review]
Excellent additions to the tutorial.

I think a mention of the Git book is obligatory: <https://git-scm.com/book/en/v2>

Most of my criticism revolves around not keeping the information on Git in general and GitHub in specific separate. However, this was already a "problem" before your changes so feel free to ignore those comments.

File: doc/GIT.md (hunk 1 above)
Comment (markus2330) [reply doc/GIT.md:156]: I agree the sentence should be corrected.

> The only dangerous thing 

As you said yourself, you lose unreachable commits after `git gc` and reflogExpireUnreachable is reached (or time jumped). So it is also not completely without any danger.

File: doc/GIT.md [hunk 2]
```diff
@@ -18,20 +29,34 @@ git config --global merge.ff false
 git config merge.ff false
```
Comment (Eiskasten) [inline doc/GIT.md:29]: Good improvement to change that to the repository only as it totally depends on the strategy of the upstream repository. In this merging section I would also add `git config pull.rebase true` which will automatically rebase when pulling which makes it unnecessary calling it manually.

File: doc/GIT.md (hunk 1 above)
Comment (Eiskasten) [reply doc/GIT.md:156]: I agree on that, but I would consider them "deleted" although as the gc can delete them unexpectedly anytime. I think you can compare the situation to java where the gc will delete objects without a reference but they remain before the gc jumps in. I think almost nobody would call them not-deleted when describing this behavior to a new user who has to understand the basics before thinking about the gc.

File: doc/GIT.md [hunk 3]
```diff
@@ -1,14 +1,25 @@
...
+git push            # Push changes to the current remote branch
 ```
 
+## Installation
+
+- [Windows](https://git-scm.com/download/win)
+- [macOS](https://git-scm.com/download/mac)
+- [Linux](https://git-scm.com/download/linux)
+
+If you are uncomfortable, with command line interfaces, there are also [plenty of GUI options available as well](https://git-scm.com/downloads/guis)
```
Comment (flo91) [inline doc/GIT.md:21]:
```suggestion
If you are uncomfortable with command line interfaces, there are also [plenty of GUI options available](https://git-scm.com/downloads/guis).
```

File: doc/GIT.md [hunk 4]
```diff
@@ -1,14 +1,25 @@
 # GIT
```
Comment (flo91) [inline doc/GIT.md:1]:
```suggestion
# Git
```
see #4679

File: doc/GIT.md [hunk 5]
```diff
@@ -1,14 +1,25 @@
...
 
-## Basic GIT Commands
+## Cheat Sheet: Basic GIT Commands
```
Comment (flo91) [inline doc/GIT.md:3]:
```suggestion
## Cheat Sheet: Basic Git Commands
```

File: doc/GIT.md [hunk 6]
```diff
@@ -18,20 +29,34 @@ git config --global merge.ff false
 git config merge.ff false
 ```
 
+This ensures that git will always create a merge commit, when you are trying to merge
```
Comment (flo91) [inline doc/GIT.md:32]:
```suggestion
This ensures that Git will always create a merge commit, when you are trying to merge.
```

File: doc/GIT.md [hunk 7]
```diff
@@ -18,20 +29,34 @@ git config --global merge.ff false
...
+Also ensure that you have your username and e-mail configured, so your work can be attributed to you:
+
+```sh
+git config --global user.name "Your Name"
+git config --global user.email "yourname@yourdomain.com"
+```
+
+### Adding/creating an SSH Key
+
+GitHub supports both HTTP and SSH for git operations. Using SSH, you can remove the annoyance of having to enter your password every time. To learn how to add an SSH-Key follow [this Guide provided by GitHub](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/adding-a-new-ssh-key-to-your-github-account)
```
Comment (flo91) [inline doc/GIT.md:43]:
```suggestion
GitHub supports both HTTP and SSH for Git operations. Using SSH, you can remove the annoyance of having to enter your password every time. To learn how to add an SSH-Key follow [this guide provided by GitHub](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/adding-a-new-ssh-key-to-your-github-account).
```

File: doc/GIT.md [hunk 8]
```diff
@@ -115,3 +140,17 @@ git push orign master:master
...
 > You don't need to switch to the `master` branch first.
 > However, this only works, if you have not modified your `master` branch, i.e. the latest commit in your forked `master` branch was once the latest commit in the main `master` branch.
+
+## Rebasing branches
+
+When you work on a local feature branch, it might happen that a change occurs within the master branch, that introduces merge conflicts. I.e. if you create a pull request, it cannot be merged into master automatically anymore. In these cases, we expect you to rebase your feature branch, so it can be merged automatically once again.
+
+To achieve this, make sure you have your master branch synced up. Then switch to the feature branch you want to update using `git checkout <feature-branch-name>` and then type `git rebase master`. Sometimes, this will result in a merge conflict.
+
+If you already have pushed changes from your feature branch to a remote branch, you'll need to either overwrite it using `git push --force` or push to a new remote branch
```
Comment (flo91) [inline doc/GIT.md:150]:
```suggestion
If you already have pushed changes from your feature branch to a remote branch, you'll need to either overwrite it using `git push --force` or push to a new remote branch.
```

File: doc/GIT.md [hunk 9]
```diff
@@ -115,3 +140,17 @@ git push orign master:master
...
+
+### Resolving merge conflicts
+
+To resolve these edit the files that need to be merged manually. You can check which files need to be merged using `git status`. After you are done merging a file manually, you can use `git add <path-to-file>` and when you have merged all files, continue with `git rebase --continue`.
```
Comment (flo91) [inline doc/GIT.md:154]:
```suggestion
To resolve merge conflicts, edit the files that need to be merged manually. You can check which files need to be merged using `git status`. After you are done merging a file manually, you can use `git add <path-to-file>` and when you have merged all files, continue with `git rebase --continue`.
```

File: doc/GIT.md (hunk 1 above)
Comment (flo91) [inline doc/GIT.md:156]:
```suggestion
> **Note**: Keep in mind that manual merges within a rebase are destructive. If you accidentally remove changes you've done in your feature branch, they will be gone forever. If you're new to manually merging files in Git, it's a good idea to create a new branch from your feature branch using `git checkout -b <new-branch-name>`. When you're stuck on a manual merge, you can also always abort the rebase using `git rebase --abort`
```

File: doc/news/_preparation_next_release.md [hunk 10]
```diff
@@ -350,6 +350,7 @@ This section keeps you up-to-date with the multi-language support provided by El
 - <<TODO>>
 - Fix grammar for `elektra-granularity.md` _(@dtdirect)_
 - Rephrase sections in doc/dev/error-\* _(@dtdirect)_
+- Improve GIT.md _(Juri Schreib @Bujuhu)_ _(Nikola Prvulovic @Dynamichost96)_
```
Comment (flo91) [inline doc/news/_preparation_next_release.md:353]:
```suggestion
- Improve Git.md _(Juri Schreib @Bujuhu)_ _(Nikola Prvulovic @Dynamichost96)_
```

**Review by flo91 — CHANGES_REQUESTED** [review]
Great improvements!
Please rebase with master and incorporate the reviews.

File: doc/GIT.md (hunk 1 above)
Comment (Bujuhu) [reply doc/GIT.md:156]: I'll try writing an explanation that's more true to life while not being too complicated

**Comment by Bujuhu** [comment]
I tried to include all reviewer suggestions. Thank you all, for your attention! I didn't put the GitHub specific features into its on section, as I don't think it makes much sense from the perspective of the reader. However, I tried to label GitHub specific features clearly

**Review by flo91 — APPROVED** [review]

**Comment by flo91** [comment]
Looks all good now, thank you!

**Pull request merged**